### ISSUE TITLE
One Controller Is Detected As Two

### ISSUE THREAD

**Issue by mermaidexe** [Ryochan7/DS4Windows#617]
Hi i'm new to GitHub, I'm using your latest release of DS4Windows. My issue is that when i'm trying to use my DS4 with Lego Star Wars for example, the second player randomly drops in. I did some digging and found that DS4Windows does in fact work correctly, but uses two instances of the controller as one. I did some more research and tried enabling "use D-input only" and it still shows two devices. I tried using "Hide DS4 controller" with no error messages, and it makes the x360 instance of the controller disappear and I cannot use the DS4 anymore after that. To my knowledge I'm not using HIDGuardian and I don't know if I'm doing anything right at all. Please reply if you can help me.

**Comment by Ryochan7** [comment]
Are you sure that you are using version 1.7.0? That version was just released yesterday so maybe you could be testing against version 1.6.14.

"Use DInput only" mode works fine for me. Using the "Hide DS4 Controller" option should only allow the virtual Xbox 360 controller to be exposed to the system unless DS4Windows failed to obtain exclusive access to the DS4. An error message will appear in the Log tab if that is the case. The DS4 controller should still be visible in DS4Windows unless exclusive mode broke and ended up permanently disabling the DS4 in Windows. There is a section of the README that covers how to check for that and what to do if the problem occurs. That type of situation is extremely rare nowadays.

https://github.com/Ryochan7/DS4Windows/blob/jay/README.md#device-detection-issue

**Comment by Ryochan7** [comment]
I should probably mention that the permanently disabling problem has not occurred to me since October 2018.

**Comment by mika-n** [comment]
Mermaid. Have you read the following Wiki page to make sure why exclusive access is needed and what this "double input/double controller" issue means? And have you followed tips in the wiki page? 
https://github.com/Ryochan7/DS4Windows/wiki/Exclusive-Mode-(Hide-DS4-Controller-config-option)-tips-and-issues

**Comment by mermaidexe** [comment]
Yes I have read both of the pages that were posted here and nothing I do is changing the outcome. I forgot to mention I am trying to use the DS4 wired, and not with bluetooth. Also yes, I am using v1.7.0 as found here: https://github.com/Ryochan7/DS4Windows/releases
Here's a screenshot of something new: https://cdn.discordapp.com/attachments/474389078402400266/558399239802060821/unknown.png
As you can see "Hide DS4Controller" is turned on with no error message and my PC is still seeing the DS4 and the x360 instance. I have nothing else open at the time of the screenshot so I know there aren't any other programs claiming the DS4.

**Comment by mika-n** [comment]
Please post a copy of the DS4Win Logs logfile option screen.

**Comment by mermaidexe** [comment]
[DS4WindowsLOG.txt](https://github.com/Ryochan7/DS4Windows/files/2994073/DS4WindowsLOG.txt)

**Comment by mika-n** [comment]
"3/21/2019 4:40:27 PM: Warning: Could not open DS4 DC:AF:68:AB:6B:D8 exclusively"

This error message in the logfile means that DS4Win app failed to acquire the exclusive access to the original DS4 controller. You should try to find out which app running in background is blocking this. Close Steam/UPlay/Netflix/GeForceExperience and so on. Open TaskManager and process list and shutdown Winstore.app.exe process and then try again to open DS4Win app and connect the controller.

**Comment by mermaidexe** [comment]
Yes I know I need to shut down all other applications that could be using the controller before DS4Windows and I have shut them all down but keep getting the same error message. Sometimes there is no error message but it still shows both instances of the controller in the devices screen. I've tried going and closing anything that could be blocking access and nothing changes. Is there something I'm missing?

**Comment by mermaidexe** [comment]
Please don't take this the wrong way. I know we've exhausted our options so I thought I would try Input Mapper's software and surprisingly, it worked the very first time I tried to use it with no tweaking to anything anywhere on my computer. I'm not saying this to try and get people to go use Input Mapper and abandon your guys' work. I'm just saying that for some reason that I can't explain, DS4 Windows failed me. I still have no clue why I couldn't get it to work. I'm sure it works great.

**Comment by mika-n** [comment]
I believe InputMapper installer installs HidGuardian tool behind the scene to hide the original controller and to expose only the virtual re-mapped output controller. I guess HidGuardian did the job to hide the original controller because it works better in scenarios where some background apps "steals" DS4 controller and prevents "exclusive mode" to do the trick without special kernel mode device driver filters (=HidGuardian device driver). 

DS4Windows can use HidGuardian also but it used to be a separate installation procedure and not very easy to do if done manually. However, Ryochan7 released a "HidGuardian Test Installer" tool to make it "one click" prosedure. For you or anyone reading this take a look at the "HidGuardian Test Installer" download link and section in Wiki pages.
https://github.com/Ryochan7/DS4Windows/wiki/DS4Windows-and-HIDGuardian-Install-and-Setup-Guide#hidguardian-test-installer

The latest DS4Windows installation also comes with "HidGuardHelper.exe" tool alongside with "DS4Windows.exe" to make certain required HidGuardian runtime configurations automagical and user doesn't have to do HidGuardian whitelisting things (ie. tells HidGuardian device driver to whitelist DS4Windows process because it should see the original DS4 controller but other processes should not).

Anyway. InputMapper is a great tool but one thing I don't like is how it went to closed source application even when it originally was a fork of open source work of ScarletCrush/JayKings and other contributors. I think InputMapper should still be an open source (it is not a problem to finance it via paid licenses and ads as long original open source license terms would not be violated and source code closed).  Even when these days lots of code in InputMapper has been re-written it's roots are nevertheless in those open source apps it originally forked. Ryochan7/DS4Windows fork is still an open source as was it's original root and never tried to hide that fact even when there is also lots of re-written code. That's why I prefer to support open source version of DS4Windows. Here is a link to history of all this "mess" with different tools and forks: http://www.myconsole.io/brief-history-playstation-controller-software/

**Comment by Ryochan7** [comment]
HidGuardian is likely being used by InputMapper in that case. DS4Windows only has minor support for HidGuardian and it requires some manual configuration to get started; there is an obscure installer that takes care of almost everything but I have not pushed for its use much. I do not know if InputMapper does something weird with its service but DS4Windows should work with the HidGuardian installation made by InputMapper. Support would likely be better if I actually used HidGuardian on a regular basis but I do not.

https://github.com/Ryochan7/DS4Windows/wiki/DS4Windows-and-HIDGuardian-Install-and-Setup-Guide#hidguardian-test-installer

InputMapper does not do its stated job right after many years in development. I have seen original versions of the flawed routines used that cause axis scaling to be wrong even in the latest 1.7 alpha client. Last time I tested, even triggers were not fully scaled and I don't remember having that problem with the 1.6 builds. During the original test of InputMapper and DS4Windows around the end of 2016, DS4Windows had better input response time so that is why I initially used it before a Windows update broke exclusive mode and I forked the project; Wobbles had posted a few IM videos on YouTube about exclusive mode breaking in IM and his decision to just rely on HidGuardian from that point.

I see myself fighting a different battle than other projects are tackling and that still makes DS4Windows relevant. If input response time were better using the Steam controller mapper, I would just use that since it is probably the most full featured mapper out there (at least for free). The inclusion of DS4 support in Steam is what got me to buy a DS4 controller in the first place; I don't own a PlayStation 4.

**Comment by Ryochan7** [comment]
It is confirmed. DS4Windows will work fine with the HidGuardian installation made by InputMapper. InputMapper installs an old version of HG v.1 (1.5.10.0) compared to the test installer (1.14.3.0) though.

**Comment by mika-n** [comment]
In my laptop the DS4Windows HidGuardian test installer (the one linked in above mentioned Wiki page) for HidGuardian installation on Win10 worked just fine and was "one click process". It installed the driver, wrote required HidG sysreg entries for DS4 controller types and voila. Done. Nice work Ryochan7. Thx for the test installer.

If there is an old version of HidGuardian already then the test installer even gives an opportunity to uninstall it. This is nice because sometimes it is a bit difficult to get rid of already installed kernel mode device drivers (if a need arise).

Also, the DS4Win supplied HidGuardHelper.exe seems to do the HidG whitelisting of a process in a simple and elegant fashion. Simple is usually the best way with computer systems.

I used DS4Windows built-in exclusive mode but my gaming laptop didn't have GeForce ingame layouts, Steam, UPlay, Origin, WinStoreApp, Netflix, Edge browser, Discord, TeamSpeak whatever tools running all the time in the background. I usually started DS4Win first and then one by one those tools I needed at the time. However, our teenage gamer has a looooong list of those background tools running in his gaming PC all the time, so HidGuardian does the job there more reliably than exclusive mode.

I know Nefarius doesn't like it when people use the existing publicly signed HidGuardian driver version, but it seems to do the job if it is used properly or tools support its API interface rules. Anyway, maybe someday the new incarnation of HidGuardian is ready to be published but for now HidG version installed with DS4Win TestInstaller is the best alternative option for those who struggle to get built-in DS4Win exclusive mode to work.

**Comment by Ryochan7** [comment]
After some recent tests with the HidGuardian 1.x source code, I think I have finally discovered a version that I could see myself using full time. The use of WdfDeviceInitSetIoType and removing the use of a WDFQUEUE, that just forwarded requests anyway, got performance to pretty much native performance. It would make sense due to removing HG from processing requests outside of CreateFile. HG performs its main job without that extra processing of requests just fine.

**Comment by mika-n** [comment]
Does this mean that the modified version should be run under "developer sign" and WinOS configuration? I guess the version of HG with a public signkey is still the "old" 1.14.3 version?

**Comment by Ryochan7** [comment]
I am using a slightly older code base than that. I have been modifying the copy that I have had since making my first pull request to HidGuardian. The reported version number is 1.13.8.1 but I think I changed the last digit to 1 so 1.13.8.0 would be a more accurate version; the Git log confirms that the changes are based on version 1.13.8.0. I need to grab a new copy of the HidGuardian code and base my changes on 1.14.3.0.

Using said changes would require running Windows in Test Mode and compiling the driver yourself. I have been messing around with driver changes so often that I have kept it enabled on my main installation. Before playing with the ViGEm drivers, vmulti was my entry into Windows driver development. I haven't used that driver in a while though.
